### PR TITLE
Monero hackathon results

### PR DESCRIPTION
**Pull request by matejcik**
This is a significant cleanup of Monero code and a good step on the way to #642.
* remove support for HF12 and below
* remove MLSAG support
* clean up monero cryptography naming
* get rid of "optional first argument" pattern, in favor of mandatory argument that is allowed to be None
  (and fix several bugs related to this feature)

One protobuf-level change is worth noting: the integer field `network_type` was replaced by an enum. This does not change the on-the-wire format.

@andrewkozlik or @onvej-sl please leave us a note whether it's ok to remove `range_proof` from the crypto library like this.

@onvej-sl you reviewed the Monero code previously so you might want to take a look; however, there are basically no functional changes here except for the removal of some steps, and the automated tests against Monero blob pass.

### CODE REVIEW

**Comment by matejcik** [comment]
@ph4r05 @mmilata @grdddj please have a look over this, whether this is what you expect to see

**Review by mmilata — APPROVED** [review]
Nice job with the typing and with the extmod. I scanned the changes and haven't spotted any issue.

File: ci/shell.nix [hunk 1]
```diff
@@ -20,7 +20,7 @@ let
   }) { };
   moneroTests = nixpkgs.fetchurl {
     url = "https://github.com/ph4r05/monero/releases/download/v0.17.3.0-dev-tests/trezor_tests";
-    sha256 = "sha256-tTQTe/Yk6oURq7GDOEotJ7Y2UpCgyuU5odjEHNTMrmE=";
+    sha256 = "b534137bf624ea8511abb183384a2d27b6365290a0cae539a1d8c41cd4ccae61";
```
[the remark this answers is not in the document]
Comment (prusnak) [reply ci/shell.nix:23]: But we can still keep the switch to hex notation as part of this PR. Honestly, I hate the base64-notation for files :-)

**Comment by grdddj** [comment]
Currently, the `unused_monero.py` snippet is not working as expected, because of file-system issues and the removal of the alias `core/src/apps/monero/xmr/crypto/__init__.py` file. I am looking at how to make it work again.

**Comment by grdddj** [comment]
> Currently, the `unused_monero.py` snippet is not working as expected, because of file-system issues and the removal of the alias `core/src/apps/monero/xmr/crypto/__init__.py` file. I am looking at how to make it work again.

Fixed in https://github.com/trezor/trezor-firmware/commit/a0ac67380c1b8da72cf1d5f7c27da8237452242c (on `grdddj/monero_hackathon_snippet_fix)` branch)

Current results say:
```
3 unused functions:
ge25519_mul8 {'mapping': ['ge25519_mul8'], 'is_used': False}
xmr_add_keys2 {'mapping': ['xmr_add_keys2'], 'is_used': False}
xmr_add_keys3 {'mapping': ['xmr_add_keys3'], 'is_used': False}
```

It really looks like these three functions from `core/mocks/generated/trezorcrypto/monero.pyi` are unused and could be deleted

**Comment by ph4r05** [comment]
Did you pls check also usage in tests? I recall mul8 might be used there

**Comment by grdddj** [comment]
> Did you pls check also usage in tests? I recall mul8 might be used there

Right, that is really used in tests - I will need to modify the script to also look there. Those two others look safe to delete, I will do it in my own branch and run CI on it

**Comment by andrewkozlik** [comment]
> @andrewkozlik or @onvej-sl please leave us a note whether it's ok to remove range_proof from the crypto library like this.

If I understand correctly, this is dead code in Trezor. In that case I am for removing it, because we won't be maintaining it. So if somebody happens to be using this function, they should fork the code and take over the maintenance.

**Comment by grdddj** [comment]
> Those two others look safe to delete, I will do it in my own branch and run CI on it

CI is green on `grdddj/monero_hackathon_snippet_fix` where I removed `xmr_add_keys2` and `xmr_add_keys3`, so we could hopefully safely merge it into this PR branch

**Comment by ph4r05** [comment]
For completeness, I got notified that a new [hardfork is being prepared](https://www.getmonero.org/2022/04/20/network-upgrade-july-2022.html), part of which is new [Bulletproof+](https://eprint.iacr.org/2020/735.pdf) protocol. I am currently studying changes required to support this protocol on Trezor. 

Fallback would be to generate it on the client-side completely, but we will need new to release a new Trezor firmware anyway to support transaction signatures on a new HF (I need to check whether it will be mandatory to use BP+ on new HF or whether we can still use old versions).

**Comment by sethforprivacy** [comment]
> I need to check whether it will be mandatory to use BP+ on new HF or whether we can still use old versions

BP+ will be mandatory after the hard-fork for all transactions, just to clarify and save you some searching :)

**Comment by ph4r05** [comment]
Note: I've implemented BP+ verifier and prover in my new branch. I also cleaned up some existing BP code. Should we add changes to this PR @matejcik ?

**Comment by grdddj** [comment]
> @grdddj please push your changes into this branch.

Done, there are four new fixups - I did not want to force-push, so they will need to be squashed

File: core/src/apps/monero/xmr/serialize/message_types.py [hunk 2]
```diff
@@ -93,42 +103,42 @@ def load(cls, reader, elem_type=None):
...
+#             if ftype.VARIANT_CODE == tag:
+#                 fvalue = ftype.load(reader)
+#                 break
+#         else:
+#             raise ValueError(f"Unknown tag: {tag}")
+#         return fvalue
+
+#     @classmethod
+#     def f_specs(cls) -> XmrFspec:
+#         return ()
```
[the remark this answers is not in the document]
Comment (matejcik) [reply core/src/apps/monero/xmr/serialize/message_types.py:138]: done in rebase

**Comment by ph4r05** [comment]
hi @matejcik , there was a change in the Monero (https://github.com/monero-project/monero/pull/8277) we might need to address by a new PR. Is there pls any codefreeze date for a new firmware release so all XMR-related PRs are merged by the next release? Thanks!

**Pull request merged**